tsl/src/compression/compression.c

```diff
@@ -1525,200 +1524,218 @@ create_per_compressed_column(TupleDesc in_desc, TupleDesc out_desc, Oid out_reli
+	 * with the count metadata column.
+	 */
+	for (int16 col = 0; col < decompressor->num_compressed_columns; col++)
+	{
+		PerCompressedColumn *column_info = &decompressor->per_compressed_cols[col];
+		if (column_info->iterator == NULL)
+		{
+			continue;
+		}
+		Assert(column_info->is_compressed);
+		const DecompressResult value = column_info->iterator->try_next(column_info->iterator);
+		CheckCompressedData(value.is_done);
+	}
+	MemoryContextSwitchTo(old_ctx);
 
-				/* Arrange for econtext's scan tuple to be the tuple under test */
-				econtext->ecxt_scantuple = tts_slot;
-#if PG14_LT
-				estate->es_result_relation_info = decompressor->indexstate;
-#endif
-				ExecInsertIndexTuplesCompat(decompressor->indexstate,
-											tts_slot,
-											estate,
-											false,
-											false,
-											NULL,
-											NIL,
-											false);
-
-				ExecDropSingleTupleTableSlot(tts_slot);
-				heap_freetuple(decompressed_tuple);
-			}
-			else
-			{
-				TupleTableSlot *slot =
-					MakeSingleTupleTableSlot(decompressor->out_desc, &TTSOpsVirtual);
-				/* create the virtual tuple slot */
-				ExecClearTuple(slot);
-				for (int i = 0; i < decompressor->out_desc->natts; i++)
-				{
-					slot->tts_isnull[i] = decompressor->decompressed_is_nulls[i];
-					slot->tts_values[i] = decompressor->decompressed_datums[i];
-				}
+	decompressor->batches_decompressed++;
+	decompressor->tuples_decompressed += n_batch_rows;
 
-				ExecStoreVirtualTuple(slot);
+	return n_batch_rows;
+}
 
-				slot_getallattrs(slot);
+void
+row_decompressor_decompress_row_to_table(RowDecompressor *decompressor)
+{
+	const int n_batch_rows = decompress_batch(decompressor);
 
-				tuplesort_puttupleslot(tuplesortstate, slot);
-				ExecDropSingleTupleTableSlot(slot);
-			}
-			wrote_data = true;
+	MemoryContext old_ctx = MemoryContextSwitchTo(decompressor->per_compressed_row_ctx);
+
+	/* Insert all decompressed rows into table using the bulk insert API. */
+	table_multi_insert(decompressor->out_rel,
+					   decompressor->decompressed_slots,
+					   n_batch_rows,
+					   decompressor->mycid,
+					   /* options = */ 0,
+					   decompressor->bistate);
+
+	/*
+	 * Now, update the indexes. If we have several indexes, we want to first
+	 * insert the entire batch into one index, then into another, and so on.
+	 * Working with one index at a time gives better data access locality,
+	 * which reduces the load on shared buffers cache.
+	 * The normal Postgres code inserts each row into all indexes, so to do it
+	 * the other way around, we create a temporary ResultRelInfo that only
+	 * references one index.
+	 */
+	ResultRelInfo single_index_rri = *decompressor->indexstate;
+	Relation single_index_relation;
+	IndexInfo *single_index_info;
+	single_index_rri.ri_NumIndices = 1;
+	single_index_rri.ri_IndexRelationDescs = &single_index_relation;
+	single_index_rri.ri_IndexRelationInfo = &single_index_info;
```

Comment:
This code looks a bit suspicious for the no index case

Reply:
There's a comment about this above -- because the PG functions want to insert each row into all indexes, we work around this by creating a temporary ResultRelInfo with only one index. It's OK even in case when there are no indexes, because the loop below won't run. I'll think if I can reorder it and improve comments to make it less suspicious. Will add a dedicated test as well for clarity, although we do have no index/many indexes tests somewhere.